runtime/src/gpu/rocm/gpu-rocm.c

```diff
@@ -27,67 +27,497 @@
 #include "chpl-tasks.h"
 #include "error.h"
 #include "chplcgfns.h"
+#include "chpl-env-gen.h"
 
 #include <assert.h>
 
-void chpl_gpu_impl_init() {}
-void chpl_gpu_impl_on_std_modules_finished_initializing() {}
+#define __HIP_PLATFORM_AMD__
+#include <hip/hip_runtime.h>
+#include <hip/hip_runtime_api.h>
+#include <hip/hip_common.h>
 
-// TODO:
-// HIP include files throw warnings when compiled under our set of
-// settings. We compile with -Werror. For the time being, do not
-// include HIP directly, and just forward declare the functions
-// ourselves.
-extern void hipGetDeviceCount(int* into);
+static void chpl_gpu_cuda_check(int err, const char* file, int line) {
+  if(err == hipErrorContextAlreadyInUse) { return; }
+  if(err != hipSuccess) {
+    const int msg_len = 256;
+    char msg[msg_len];
+
+    snprintf(msg, msg_len,
+             "%s:%d: Error calling HIP function: %s (Code: %d)",
+             file, line, hipGetErrorString((hipError_t)err), err);
+
+    chpl_internal_error(msg);
+  }
+}
+
+#define CUDA_CALL(call) do {\
+  chpl_gpu_cuda_check((int)call, __FILE__, __LINE__);\
+} while(0);
+
+static inline
+void* chpl_gpu_load_module(const char* fatbin_data) {
+  hipModule_t cuda_module;
+
+  CUDA_CALL(hipModuleLoadData(&cuda_module, fatbin_data));
+  assert(cuda_module);
+
+  return (void*)cuda_module;
+}
+
+static inline
+void* chpl_gpu_load_function(hipModule_t cuda_module, const char* kernel_name) {
```

Comment:
A search and replace for `cuda` and `CUDA` is needed in this file :)

Reply:
I've done: `s/cuda/rocm/g` and `s/CUDA/ROCM/g`, but if we do end up using hipify to autogenerate this file I think we'll want to come up with more generic names in our CUDA version (or whatever version we consider the "original") as well.